### PR TITLE
Enable support for cython 3

### LINKED ISSUE #211: Does not build with Cython 3
Cython 3.0.0 was just released (https://cython.readthedocs.io/en/latest/src/changes.html), and it looks like `cyipopt` will need some changes to be compatible with it.

```
python3 -m build
```

```
[...]
Compiling cyipopt/cython/ipopt_wrapper.pyx because it changed.
[1/1] Cythonizing cyipopt/cython/ipopt_wrapper.pyx
/tmp/build-env-9tlwx0y5/lib64/python3.11/site-packages/Cython/Compiler/Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/build-via-sdist-_jz2qe4_/cyipopt-1.2.0.dev0/cyipopt/cython/ipopt_wrapper.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)

Error compiling Cython file:
------------------------------------------------------------
...
                                        <Number*>np_cl.data,
                                        <Number*>np_cu.data,
                                        nele_jac,
                                        nele_hess,
                                        0,
                                        objective_cb,
                                        ^
------------------------------------------------------------

cyipopt/cython/ipopt_wrapper.pyx:416:40: Cannot assign type 'Bool (Index, Number *, Bool, Number *, UserDataPtr) except? -1' to 'Eval_F_CB'

Error compiling Cython file:
------------------------------------------------------------
...
                                        <Number*>np_cu.data,
                                        nele_jac,
                                        nele_hess,
                                        0,
                                        objective_cb,
                                        constraints_cb,
                                        ^
------------------------------------------------------------

cyipopt/cython/ipopt_wrapper.pyx:417:40: Cannot assign type 'Bool (Index, Number *, Bool, Index, Number *, UserDataPtr) except? -1' to 'Eval_G_CB'

Error compiling Cython file:
------------------------------------------------------------
...
                                        nele_jac,
                                        nele_hess,
                                        0,
                                        objective_cb,
                                        constraints_cb,
                                        gradient_cb,
                                        ^
------------------------------------------------------------

cyipopt/cython/ipopt_wrapper.pyx:418:40: Cannot assign type 'Bool (Index, Number *, Bool, Number *, UserDataPtr) except? -1' to 'Eval_Grad_F_CB'

Error compiling Cython file:
------------------------------------------------------------
...
                                        nele_hess,
                                        0,
                                        objective_cb,
                                        constraints_cb,
                                        gradient_cb,
                                        jacobian_cb,
                                        ^
------------------------------------------------------------

cyipopt/cython/ipopt_wrapper.pyx:419:40: Cannot assign type 'Bool (Index, Number *, Bool, Index, Index, Index *, Index *, Number *, UserDataPtr) except? -1' to 'Eval_Jac_G_CB'

Error compiling Cython file:
------------------------------------------------------------
...
                                        0,
                                        objective_cb,
                                        constraints_cb,
                                        gradient_cb,
                                        jacobian_cb,
                                        hessian_cb
                                        ^
------------------------------------------------------------

cyipopt/cython/ipopt_wrapper.pyx:420:40: Cannot assign type 'Bool (Index, Number *, Bool, Number, Index, Number *, Bool, Index, Index *, Index *, Number *, UserDataPtr) except? -1' to 'Eval_H_CB'

Error compiling Cython file:
------------------------------------------------------------
...

        if self.__nlp == NULL:
            msg = "Failed to create NLP problem. Make sure inputs are okay!"
            raise RuntimeError(msg)

        SetIntermediateCallback(self.__nlp, intermediate_cb)
                                            ^
------------------------------------------------------------

cyipopt/cython/ipopt_wrapper.pyx:427:44: Cannot assign type 'Bool (Index, Index, Number, Number, Number, Number, Number, Number, Number, Number, Index, UserDataPtr) except? -1' to 'Intermediate_CB'
Traceback (most recent call last):
[...]
```

### PR DESCRIPTION
**Pull request by chrhansk**
This is an attempt to fix #211:

- All callbacks are now marked `noexpect` with exceptions being handled (stored) internally
- A large `try` / `expect` block is used in the callbacks, where code outside of the blocks should not throw (only stack allocations)
- The large Jacobian / Hessian callbacks have been split up into value / structure callbacks
- Cython restriction to <3 was removed

### CODE REVIEW

**Comment by chrhansk** [comment]
Edit: Everything is cast to `Problem` rather than `object` in this patch, so the callbacks have access to all (not just public) `cdef` properties, which could now be made private (by removing the `public`) without problems.

**Comment by moorepants** [comment]
This will need to run with Cython 0.29.* and 3+. Does it?

**Comment by chrhansk** [comment]
It is supposed to and does on my system.

**Comment by musicinmybrain** [comment]
> This will need to run with Cython 0.29.* and 3+. Does it?

This works for me (builds and tests pass, except any pre-existing issues like https://github.com/mechmotum/cyipopt/issues/237 and https://github.com/mechmotum/cyipopt/issues/238) on Fedora Linux Rawhide as well. I’ll probably go ahead and apply this as a downstream patch in Rawhide, since this is one of a handful of packages that still needs Cython 0.29, and we’d like to stop maintaining that compat package sooner rather than later.

File: cyipopt/cython/ipopt_wrapper.pyx [hunk 1]
```diff
@@ -858,16 +858,20 @@ cdef Bool objective_cb(Index n,
...
+    cdef np.ndarray[DTYPEd_t, ndim=1] _x
+
     try:
+        log(b"objective_cb", logging.INFO)
+
+        self = <Problem>user_data
+        _x = np.zeros((n,), dtype=DTYPEd)
+
+        for i in range(n):
+            _x[i] = x[i]
```
Comment (moorepants) [inline cyipopt/cython/ipopt_wrapper.pyx:873]: How will these new lines raise a `CyIpoptEvalutionError`? It isn't clear why you've moved them in this try/except.

**Comment by moorepants** [comment]
I've gone through this and I don't see any issues with the changes. Thanks.

**Comment by chrhansk** [comment]
Thank you for the merge.

**Pull request merged**